lib/Wikia/tests/Util/SamplerProxyTest.php

```diff
@@ -0,0 +1,312 @@
+			$samplerProxy->getMethodSamplingRateVariableName() );
+		$this->assertEquals( $testOriginalCallable, $samplerProxy->getOriginalCallable() );
+		$this->assertEquals( $testAlternateCallable, $samplerProxy->getAlternateCallable() );
+		$this->assertEquals( $testResultsCallable, $samplerProxy->getResultsCallable() );
+	}
+
+	function testUnsampledMethod() {
+		$testShadowVariableName = 'wgEnableShadowing';
+		$GLOBALS[$testShadowVariableName] = 0;
+		$testMethodSamplingRateVariableName = 'wgMethodSamplingRate';
+		$GLOBALS[$testMethodSamplingRateVariableName] = 100;
+
+		$testOriginalCallable = [ $this->originalMock, 'methodToSample' ];
+		$testAlternateCallable = [ $this->alternateMock, 'alternateMethod' ];
+
+		$testArg1 = 1;
+		$testArg2 = array( 4, 5, 6 );
+		$testResult = 'success';
+
+		$builder = SamplerProxy::createBuilder();
+		$samplerProxy =
+			$builder->enableShadowingVariableName( $testShadowVariableName )
+				->methodSamplingRateVariableName( $testMethodSamplingRateVariableName )
+				->originalCallable( $testOriginalCallable )
+				->alternateCallable( $testAlternateCallable )
+				->build();
+
+		$this->originalMock->expects( $this->once() )
+			->method( 'methodNotToSample' )
+			->with( $testArg1, $testArg2 )
+			->willReturn( $testResult );
+
+		$result = $samplerProxy->methodNotToSample( $testArg1, $testArg2 );
+
+		$this->assertEquals( $testResult, $result );
+	}
+
+	function testSampling100Percent() {
+		$this->doSamplingTest( 100, false, 200, 0.0 );
+	}
+
+	function testSampling75Percent() {
+		$this->doSamplingTest( 75, false, 200, 0.1 );
+	}
+
+	function testSampling50Percent() {
+		$this->doSamplingTest( 50, false, 200, 0.1 );
+	}
+
+	function testSampling25Percent() {
+		$this->doSamplingTest( 25, false, 200, 0.1 );
+	}
+
+	function testSampling0Percent() {
+		$this->doSamplingTest( 0, false, 200, 0 );
+	}
+
+	function testShadowing100Percent() {
+		$this->doSamplingTest( 100, true, 200, 0.0 );
+	}
+
+	function testShadowing75Percent() {
+		$this->doSamplingTest( 75, true, 200, 0.1 );
+	}
+
+	function testShadowing50Percent() {
+		$this->doSamplingTest( 50, true, 200, 0.1 );
+	}
+
+	function testShadowing25Percent() {
+		$this->doSamplingTest( 25, true, 200, 0.1 );
+	}
+
+	function testShadowing0Percent() {
+		$this->doSamplingTest( 0, true, 200, 0 );
+	}
+
+	function doSamplingTest(
+		$samplingRate, $enableShadowing, $count, $deviation
+	) {
+
+		$testShadowVariableName = 'wgEnableShadowing';
+		$GLOBALS[$testShadowVariableName] = $enableShadowing;
+		$testMethodSamplingRateVariableName = 'wgMethodSamplingRate';
+		$GLOBALS[$testMethodSamplingRateVariableName] = $samplingRate;
+
+		$testOriginalCallable = [ $this->originalMock, 'methodToSample' ];
+		$testAlternateCallable = [ $this->alternateMock, 'alternateMethod' ];
+
+		$testArg1 = 1;
+		$testArg2 = 'two';
+		$testArg3 = array( 4, 5, 6 );
+		$originalTestResult = 'original';
+		$alternateTestResult = 'alternate';
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Define a constant instead of duplicating this literal "alternate" 3 times. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1192)